### PR TITLE
Decompile ST/NO3 EntityStrongWarg

### PR DESCRIPTION
**Pull request by Max-Clark**
Required code moving @ padding for jumptable to align (thanks @sozud !). The new function is in [4CA54.c](https://github.com/Xeeynamo/sotn-decomp/compare/master...Max-Clark:sotn-decomp:EntityStrongWarg?expand=1#diff-0c3d267f1984dee80517bbcbafd22a0774290aed49f88dd5f038d8eb634eb0ae)

### CODE REVIEW

File: src/st/no3/4CA54.c [hunk 1]
```diff
@@ -0,0 +1,1068 @@
...
+        }
+    }
+
+    var_s1 = self->step;
+    switch (var_s1) {
+    case 0:
+        ent_s0 = self + 1;
+        self->unk60 = ent_s0;
+        // PSP version: 0x20
+        CreateEntityFromCurrentEntity(0x30, ent_s0);
```
Comment (JoshSchreuder) [inline src/st/no3/4CA54.c:180]: `E_ID_30` https://github.com/Xeeynamo/sotn-decomp/blob/95930a2fe303e0e00063af80ff8c07421cda7bbd/src/st/no3/no3.h#L50

You could also add `E_ID_31` for line 189 below

File: src/st/no3/4CA54.c [hunk 2]
```diff
@@ -0,0 +1,1068 @@
...
+            CreateEntityFromCurrentEntity(E_EXPLOSION, newEntity);
+            point = &D_80183060[(Random() & 7)];
+            newEntity->posX.i.hi += point->x;
+            newEntity->posY.i.hi += point->y;
+            newEntity->params = 1;
+        }
+    }
+    entity->ext.entityExplosion3.timer++;
+    if (entity->ext.entityExplosion3.timer >= 32) {
+        CreateEntityFromCurrentEntity(2, entity);
```
Comment (JoshSchreuder) [inline src/st/no3/4CA54.c:1007]: `E_EXPLOSION`

File: src/st/no3/4CA54.c [hunk 3]
```diff
@@ -0,0 +1,1068 @@
...
+        if (!(self->ext.strongWarg.unk7C & 0xFF)) {
+            if (self->params) {
+                frameIdx = AnimateEntity(&D_801827EC, self);
+            } else {
+                frameIdx = AnimateEntity(&D_801827DC, self);
+            }
+            if (!frameIdx || frameIdx & 0x80) {
+                frameIdx = (self->animFrameIdx - 1);
+                if (self->facingLeft) {
+                    self->velocityX = *(&D_801829EC + frameIdx);
```
Comment (sozud) [inline src/st/no3/4CA54.c:238]: Does this work if `D_801829EC[frameIdx]` ? same for other array accesses

File: src/st/no3/4CA54.c [hunk 4]
```diff
@@ -0,0 +1,1068 @@
...
+            if (self->facingLeft) {
+                g_api.CheckCollision(
+                    self->posX.i.hi - 0x38, self->posY.i.hi, &collider, 0);
+            } else {
+                g_api.CheckCollision(
+                    self->posX.i.hi + 0x38, self->posY.i.hi, &collider, 0);
+            }
+
+            if ((!self->ext.factory.unk86) ||
+                (collider.effects & 1 && collider.effects & 2)) {
```
Comment (sozud) [inline src/st/no3/4CA54.c:295]: enum https://github.com/Xeeynamo/sotn-decomp/blob/c228c6f26483c4970b7f2d0f11a300fa4ee67f56/include/game.h#L1277

File: src/st/no3/4CA54.c [hunk 5]
```diff
@@ -0,0 +1,1068 @@
...
+            }
+
+            if ((var_s1 & 0x80) && (ent_s0->animFrameIdx == 5)) {
+                if (self->facingLeft) {
+                    self->velocityX = FIX(4.0);
+                } else {
+                    self->velocityX = FIX(-4.0);
+                }
+
+                PlaySfxPositional(0x6C7);
```
Comment (sozud) [inline src/st/no3/4CA54.c:629]: My counting may be wrong but I think we have an enum https://github.com/Xeeynamo/sotn-decomp/blob/c228c6f26483c4970b7f2d0f11a300fa4ee67f56/include/sfx.h#L502

**Comment by sozud** [comment]
This got conflicted and needs a rebase

File: src/st/no3/4CA54.c [hunk 6]
```diff
@@ -0,0 +1,1068 @@
...
+            AnimateEntity(&D_8018281C, self);
+        }
+        if ((GetDistanceToPlayerX() < 0x80) || (self->hitFlags)) {
+            SetStep(3);
+        }
+        break;
+    case 3:
+        func_801CC90C(self);
+        var_s1 = GetDistanceToPlayerX();
+        if (self->ext.factory.unk80) {
```
[the remark this answers is not in the document]
Comment (Max-Clark) [reply src/st/no3/4CA54.c]: There are some cases where 0x7C is accessed as a u8 (e.g., `ext.stub[0]`) and sometimes as an s32. I have the u8 accesses in as ext.stub[0], is that alright or should I create a union in the strongwarg entity typedef (or am I missing something else?)

File: src/st/no3/4CA54.c [hunk 7]
```diff
@@ -0,0 +1,1070 @@
...
+                frameIdx = self->animFrameIdx - 1;
+                if (self->facingLeft) {
+                    self->velocityX = -D_80182A04[frameIdx];
+                } else {
+                    self->velocityX = D_80182A04[frameIdx];
+                }
+            }
+
+            if (var_s1 >= 0x79) {
+                self->ext.stub[0] = 0;
```
Comment (sozud) [inline src/st/no3/4CA54.c:265]: Fix both `ext.stub` to be strongWarg and I think this is good to go

File: src/st/no3/4CA54.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (sozud) [reply src/st/no3/4CA54.c]: We have some premade union types if any of those work. I think a union is fine for now https://github.com/Xeeynamo/sotn-decomp/blob/d9f9d9c79c393c17207e9660de7698065da0416c/include/entity.h#L11

**Pull request merged**